infra/base-images/base-clang-introspector/checkout_build_install_llvm.sh

```diff
@@ -0,0 +1,259 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+
+# See issue #4270. The compiler crashes on GCB instance with 32 vCPUs, so when
+# we compile on GCB we want 16 cores. But locally we want more (so use nproc /
+# 2).
+NPROC=$(expr $(nproc) / 2)
+
+# zlib1g-dev is needed for llvm-profdata to handle coverage data from rust compiler
+LLVM_DEP_PACKAGES="build-essential make cmake ninja-build git python3 python3-distutils g++-multilib binutils-dev zlib1g-dev"
+apt-get update && apt-get install -y $LLVM_DEP_PACKAGES --no-install-recommends
+
+# Checkout
+CHECKOUT_RETRIES=10
+function clone_with_retries {
+  REPOSITORY=$1
+  LOCAL_PATH=$2
+  CHECKOUT_RETURN_CODE=1
+
+  # Disable exit on error since we might encounter some failures while retrying.
+  set +e
+  for i in $(seq 1 $CHECKOUT_RETRIES); do
+    rm -rf $LOCAL_PATH
+    git clone $REPOSITORY $LOCAL_PATH
+    CHECKOUT_RETURN_CODE=$?
+    if [ $CHECKOUT_RETURN_CODE -eq 0 ]; then
+      break
+    fi
+  done
+
+  # Re-enable exit on error. If checkout failed, script will exit.
+  set -e
+  return $CHECKOUT_RETURN_CODE
+}
+
+function cmake_llvm {
+  extra_args="$@"
+  cmake -G "Ninja" \
+      -DLIBCXX_ENABLE_SHARED=OFF \
+      -DLIBCXX_ENABLE_STATIC_ABI_LIBRARY=ON \
+      -DLIBCXXABI_ENABLE_SHARED=OFF \
+      -DCMAKE_BUILD_TYPE=Release \
+      -DLLVM_TARGETS_TO_BUILD="$TARGET_TO_BUILD" \
+      -DLLVM_ENABLE_PROJECTS="$PROJECTS_TO_BUILD" \
+      -DLLVM_BINUTILS_INCDIR="/usr/include/" \
+      $extra_args \
+      $LLVM_SRC/llvm
+}
+
+apt-get install -y texinfo bison flex
+
+# Use chromium's clang revision
+mkdir $SRC/chromium_tools
+cd $SRC/chromium_tools
+git clone https://chromium.googlesource.com/chromium/src/tools/clang --depth 1
+cd clang
+
+LLVM_SRC=$SRC/llvm-project
+
+# For manual bumping.
+OUR_LLVM_REVISION=llvmorg-14-init-7378-gaee49255
+
+# To allow for manual downgrades. Set to 0 to use Chrome's clang version (i.e.
+# *not* force a manual downgrade). Set to 1 to force a manual downgrade.
+FORCE_OUR_REVISION=1
+LLVM_REVISION=$(grep -Po "CLANG_REVISION = '\K([^']+)" scripts/update.py)
+
+clone_with_retries https://github.com/llvm/llvm-project.git $LLVM_SRC
+
+set +e
+git -C $LLVM_SRC merge-base --is-ancestor $OUR_LLVM_REVISION $LLVM_REVISION
+IS_OUR_REVISION_ANCESTOR_RETCODE=$?
+set -e
+
+# Use our revision if specified by FORCE_OUR_REVISION or if our revision is a
+# later revision than Chrome's (i.e. not an ancestor of Chrome's).
+if [ $IS_OUR_REVISION_ANCESTOR_RETCODE -ne 0 ] || [ $FORCE_OUR_REVISION -eq 1 ] ; then
+  LLVM_REVISION=$OUR_LLVM_REVISION
+fi
+
+git -C $LLVM_SRC checkout $LLVM_REVISION
+echo "Using LLVM revision: $LLVM_REVISION"
+
+### For fuzz introspector
+echo "Applying introspector changes"
```

Comment:
To avoid lots of code duplication, we can edit this file in `base-clang` instead.

We can just guard this with:

```bash
if [ -n "$INTROSPECTOR_PATCHES ]; then
 ...
fi
```

Then we can get rid of the `base-clang-introspector` dir completely ,and modify `base-clang`s Dockerfile to do something like this using a [runtime build arg](https://docs.docker.com/engine/reference/builder/#arg).

```dockerfile
ARG introspector
ENV INTROSPECTOR_PATCHES=$instrospector
```

So that we can use the same Dockerfile to build both vanilla clang and patched clang:

```
docker build -t gcr.io/oss-fuzz-base/base-clang infra/base-images/base-clang  # vanilla
docker build --build-arg introspector=1 -t gcr.io/oss-fuzz-base/base-clang:introspector infra/base-images/base-clang   # introspector
```

Reply:
Done.